selfdrive/car/card.py

```diff
@@ -58,8 +61,74 @@ def can_send(msgs: list[CanData]) -> None:
   return can_recv, can_send
 
 
+def is_dataclass(obj):
+  """Similar to dataclasses.is_dataclass without instance type check checking"""
+  return hasattr(obj, _FIELDS)
+
+
+def asdictref(obj) -> dict[str, Any]:
+  """
+  Similar to dataclasses.asdict without recursive type checking and copy.deepcopy
+  Note that the resulting dict will contain references to the original struct as a result
+  """
+  if not is_dataclass(obj):
+    raise TypeError("asdictref() should be called on dataclass instances")
+
+  def _asdictref_inner(obj) -> dict[str, Any] | Any:
+    if is_dataclass(obj):
+      ret = {}
+      for field in getattr(obj, _FIELDS):  # similar to dataclasses.fields()
+        ret[field] = _asdictref_inner(getattr(obj, field))
+      return ret
+    elif isinstance(obj, (tuple, list)):
+      return type(obj)(_asdictref_inner(v) for v in obj)
+    else:
+      return obj
+
+  return _asdictref_inner(obj)
```

Comment:
We unfortunately have to rewrite these to take advantage of referencing vs. deepcopying and avoiding calling `type()` on every single field recursively, which take around ~7-8% total vs 3-4% with this